src/main/scala/temple/DSL/semantics/Analyser.scala

```diff
@@ -0,0 +1,395 @@
+      * @tparam A the type of metadata parsed by `parser`, which must be equal to/a subtype of this parser’s metadata
+      *           type
+      */
+    def inheritFrom[A <: T](parser: MetadataParser[A]): Unit =
+      // insert, or no-op if they already exist as inheritance does not overwrite newly added values
+      parser.matchers.foreach(matchers.safeInsert(_, ()))
+
+    /** Used for a fluent API style: `inherit from myMetadataParser` as an alternative to
+      * `inheritFrom(myMetadataParser)` */
+    final protected object inherit {
+
+      /** An alias of [[MetadataParser#inheritFrom]], for a fluent API */
+      def from[A <: T](parser: MetadataParser[A]): Unit = inheritFrom(parser)
+    }
+
+    /**
+      * Add a handler for a new type of metadata
+      *
+      * @param metaKey The name of the metadata item to add
+      * @param argKey The name of the single argument to the metadata. Note that there is also
+      *               [[temple.DSL.semantics.Analyser.MetadataParser#registerKeyword]]
+      * @param argType The type of the field to expect
+      * @param constructor The function to turn an input of type [[ArgType]] into a value of type [[T]]
+      * @tparam A The underlying type of the field, inferred from `argType`
+      */
+    // TODO: do we need to add support for multiple arguments in future?
+    final protected def registerKeyword[A](metaKey: String, argKey: String, argType: ArgType[A])(
+      constructor: A => T
+    ): Unit =
+      matchers += (metaKey -> { args =>
+          implicit val context: Context = Context(metaKey)
+          val argMap                    = parseParameters(argKey -> None)(args)
+          constructor(argMap.getArg(argKey, argType))
+        })
+
+    /** A shorthand for [[temple.DSL.semantics.Analyser.MetadataParser#registerKeyword]] with the same `key` used for both the
+      * metadata name and its single argument */
+    final protected def registerKeyword[A](key: String, argType: ArgType[A])(constructor: A => T): Unit =
+      registerKeyword(key, key, argType)(constructor)
+
+    /** Perform parsing by looking up the relevant function and
+      * [[scala.collection.IterableOnceOps#foldRight(java.lang.Object, scala.Function2)]] supports */
+    final def apply(metaKey: String, args: Args)(implicit context: BlockContext): T =
+      matchers.get(metaKey).map(_(args)) getOrElse {
+        fail(s"No valid metadata $metaKey in $context")
+      }
+  }
+
+  /** A parser of Metadata items that can occur in all the blocks */
+  private val parseGlobalMetadata = new MetadataParser[GlobalMetadata] {
+    registerKeyword[String]("language", TokenArgType)(Language.parse(_))
+  }
+
+  /** A parser of Metadata items that can occur in project and service blocks */
+  private val parseProjectAndServiceMetadata = new MetadataParser[ProjectAndServiceMetadata] {
+    inherit from parseGlobalMetadata
+    registerKeyword("database", TokenArgType)(Database.parse(_))
+  }
```

Comment:
I can't add this to a constructor because of the additional type parameters, but thoughts on this (versus `inheritFrom(parseGlobalMetadata)`, or another way of expressing it/building up these parsers?

Reply:
Agree with Jay on just explicitly starting the complete list of keywords at each bottom level object and taking the repetition hit